platooning_control_ihp/src/platoon_control_ihp_worker.cpp

```diff
@@ -45,6 +45,74 @@ namespace platoon_control_ihp
 		current_pose_ = msg.pose;
 	}
 
+    // UCLA: Put the IHP gap regulation at low level control plug-in
+    double PlatoonControlIHPWorker::getIHPDesPosFollower(double leaderCurrentPosition)
+    {
+        /**
+         * Calculate desired position based on previous vehicle's trajectory for followers.
+         * 
+         * TODO: The platoon trajectory regulation is derived with the assumption that all vehicle 
+         *       have identical length (i.e., 5m). Future development is needed to include variable 
+         *       vehicle length into consideration.
+         */
+
+
+        // 1. find the summation of "veh_len/veh_speed" for all predecessors
+        double tmp_time_hdw = current_predecessor_time_headway_sum_;
+
+        // 2. read host veh and front veh info 
+        // Predecessor vehicle data.
+        double pred_spd = predecessor_speed_;       // m/s 
+        double pred_pos = predecessor_position_;    // m
+        
+        // host data. 
+        double ego_spd = currentSpeed;                          // m/s
+        double ego_pos = leaderCurrentPosition - actual_gap_;   // m
+        
+        // platoon position index
+        int pos_idx = host_platoon_position_;
+
+        double desirePlatoonGap = ctrl_config_.intra_tau;       //s
+        
+        // IHP desired position calculation methods
+        double pos_g;      // desired downtrack position calculated with time-gap, in m.
+        double pos_h;      // desired downtrack position calculated with distance headaway, in m.
+
+        // 3. IHP gap regualtion 
+        // intermediate variables 
+        double timeGapAndStepRatio = desirePlatoonGap/ctrl_config_.time_step;     // The ratio between desired platoon time gap and the current time_step.
+        double totalTimeGap = desirePlatoonGap*pos_idx;                           // The overall time gap from host vehicle to the platoon leader, in s.
+
+        // calcilate pos_gap and pos_headway
+        if ((pred_spd <= ego_spd) && (ego_spd <= ctrl_config_.ss_theta))
+        {
+            // pos_g 
+            pos_g = (pred_pos - ctrl_config_.vehicleLength - ctrl_config_.standstill + ego_pos*timeGapAndStepRatio) / (1 + timeGapAndStepRatio);
+            // pos_h
+            double pos_h_nom = (pred_pos - ctrl_config_.standstill + ego_pos*(totalTimeGap + tmp_time_hdw)/ctrl_config_.time_step);
```

Comment:
I'm concerned about use of tmp_time_hdw here, since I believe it represents all vehicles between leader and host; however, this equation is using pred_pos, which only represents the immediate predecessor.  It seems this is a mixture of gap control and headway control.  Also, it seems the sum totalTimeGap + tmp_time_hdw is sort of double-counting the time gap between leader and host.  If this is correct, then a comment explaining it would be valuable, since, again, this sum doesn't appear in the design doc.
This comment applies to code in the else block below, also.